File: HNW_validation/HNW_validation_helper.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import LogNorm
import os
import pandas as pd
import xarray as xr





def _filter_season(df, full_season=False):
    """Filter DataFrame to snow season (Nov 1 – Apr 30) unless full_season=True."""
    
    df = _set_datime_index(df)

    if full_season:
        return df.copy()
    
    else: 
        mask = (df.index.month >= 11) | (df.index.month <= 4)
        return df.loc[mask].copy()

def _set_datime_index(df):
    """Ensure DataFrame has a DatetimeIndex."""
    
    df.index = pd.to_datetime(df.time)
    return df


def _calculate_metrics(x, y):
    """
    Calculate validation metrics.
    x = observed
    y = modeled
    """

    residuals = y - x

    rmse = np.sqrt(np.mean(residuals**2))
    bias = np.mean(residuals)
    pbias = np.sum(residuals) / np.sum(x) if np.sum(x) != 0 else np.nan

    ss_res = np.sum((x - y)**2)
    ss_tot = np.sum((x - np.mean(x))**2)
    r2 = 1 - ss_res / ss_tot if ss_tot != 0 else np.nan

    return {
        "RMSE": rmse,
        "Bias": bias,
        "Rel_BIAS": pbias,
        "R2": r2
    }


def _plot_validation(x, y, stats, model_name, lim, xlabel, ylabel):
    """
    Generic validation density plot.
    """

    vmax = max(1, len(x) / 10)

    plt.figure(figsize=(8, 7))

    plt.hist2d(
        x, y,
        bins=50,
        range=[lim, lim],
        norm=LogNorm(vmin=1, vmax=vmax),
        cmap="viridis"
    )

    cb = plt.colorbar(label="Number of observations")

    ticks = [t for t in [1, 10, 100, 1000, 10000] if t <= vmax]
    if vmax not in ticks:
        ticks.append(vmax)

    cb.set_ticks(ticks)
    cb.set_ticklabels([f"{int(t)}" for t in ticks[:-1]] + [f"{int(vmax)}"])

    plt.plot(lim, lim, "--", color="gray", linewidth=1.3)

    ticks_xy = np.linspace(lim[0], lim[1], 5)
    plt.xticks(ticks_xy)
    plt.yticks(ticks_xy)

    plt.xlabel(xlabel)
    plt.ylabel(ylabel)
    plt.title(model_name, fontsize=15)

    textstr = (
        f"$R^2$: {stats['R2']:.2f}\n"
        f"Bias: {stats['Bias']:.2f}\n"
        f"RMSE: {stats['RMSE']:.1f}\n"
        f"Rel_BIAS: {stats['Rel_BIAS']:.1%}\n"
    )

    plt.text(
        0.03, 0.97, textstr,
        transform=plt.gca().transAxes,
        fontsize=12,
        verticalalignment="top",
        bbox=dict(boxstyle="round", facecolor="white", alpha=0.8)
    )

    plt.xlim(lim)
    plt.ylim(lim)

    plt.grid(False)
    plt.tight_layout()


def _plot_validation_ax(ax, x, y, stats, title, lim, xlabel, ylabel, fig):
    """Draw a density validation plot into an existing Axes."""

    vmax = max(1, len(x) / 10)

    h, xedges, yedges, img = ax.hist2d(
        x, y,
        bins=50,
        range=[lim, lim],
        norm=LogNorm(vmin=1, vmax=vmax),
        cmap="viridis"
    )

    cb = fig.colorbar(img, ax=ax, label="Number of observations")

    ticks = [t for t in [1, 10, 100, 1000, 10000] if t <= vmax]
    if vmax not in ticks:
        ticks.append(vmax)

    cb.set_ticks(ticks)
    cb.set_ticklabels([f"{int(t)}" for t in ticks[:-1]] + [f"{int(vmax)}"])

    ax.plot(lim, lim, "--", color="gray", linewidth=1.3)

    ticks_xy = np.linspace(lim[0], lim[1], 5)
    ax.set_xticks(ticks_xy)
    ax.set_yticks(ticks_xy)

    ax.set_xlabel(xlabel)
    ax.set_ylabel(ylabel)
    ax.set_title(title, fontsize=13)

    textstr = (
        f"$R^2$: {stats['R2']:.2f}\n"
        f"Bias: {stats['Bias']:.2f}\n"
        f"RMSE: {stats['RMSE']:.1f}\n"
        f"Rel_BIAS: {stats['Rel_BIAS']:.1%}\n"
        f"N: {stats['N']}"
    )

    ax.text(
        0.03, 0.97, textstr,
        transform=ax.transAxes,
        fontsize=11,
        verticalalignment="top",
        bbox=dict(boxstyle="round", facecolor="white", alpha=0.8)
    )

    ax.set_xlim(lim)
    ax.set_ylim(lim)
    ax.grid(False)


def validate_hnw_mag25(df,
                       model_name,
                       obs_col="HNW_obs",
                       mod_col="HNW_mod",
                       save_dir=None,
                       filename="hnw_validation.png",
                       full_season=False,
                       drop_weisfluh_joch=True, ax=None):
    if drop_weisfluh_joch:
        df = df[df["station"] != "Weisfluh_Joch"].copy()

    df = _filter_season(df, full_season)

    df_valid = df.dropna(subset=[obs_col, mod_col])

    

    # Caution greate differnce if it is >= 0 or > 0
    df_valid = df_valid[df_valid[obs_col] >= 0]

    df_valid = df_valid[
        np.isfinite(df_valid[obs_col]) &
        np.isfinite(df_valid[mod_col])
    ]

    x = df_valid[obs_col].values
    y = df_valid[mod_col].values

    stats = _calculate_metrics(x, y)
    stats["N"] = len(df_valid)

    print(stats)

    _plot_validation(
        x,
        y,
        stats,
        model_name,
        lim=[0, 100],
        xlabel="Observed HNW (mm)",
        ylabel="Modeled HNW (mm)"
    )

    if save_dir is not None:
        os.makedirs(save_dir, exist_ok=True)
        save_path = os.path.join(save_dir, filename)
        plt.savefig(save_path, dpi=300, bbox_inches="tight")
        print(f"Plot saved to: {save_path}")

    plt.show()

    return stats


def validate_swe_mag25(df,
                       model_name,
                       obs_col="SWE_obs",
                       mod_col="SWE_mod",
                       save_dir=None,
                       filename="swe_validation.png",
                       full_season=False,
                       drop_weisfluh_joch=True):


    if drop_weisfluh_joch:
        df = df[df["station"] != "Weisfluh_Joch"].copy()

    df = _filter_season(df, full_season)

    df_valid = df.dropna(subset=[obs_col, mod_col])

    df_valid = df_valid[df_valid[obs_col] >= 0]

    print(f"Number of valid observations after filtering: {len(df_valid)}")

    df_valid = df_valid[
        np.isfinite(df_valid[obs_col]) &
        np.isfinite(df_valid[mod_col])
    ]

    x = df_valid[obs_col].values
    y = df_valid[mod_col].values

    stats = _calculate_metrics(x, y)
    stats["N"] = len(df_valid)

    print(stats)

    _plot_validation(
        x,
        y,
        stats,
        model_name,
        lim=[0, 1000],
        xlabel="Observed SWE (mm)",
        ylabel="Modeled SWE (mm)"
    )

    if save_dir is not None:
        os.makedirs(save_dir, exist_ok=True)
        save_path = os.path.join(save_dir, filename)
        plt.savefig(save_path, dpi=300, bbox_inches="tight")
        print(f"Plot saved to: {save_path}")

    plt.show()

    return stats


def validate_hnw_swe_combined(hnw_df, swe_df, model_name,
                               params=None,
                               hnw_obs_col="HNW_obs", hnw_mod_col="HNW_mod",
                               swe_obs_col="SWE_obs", swe_mod_col="SWE_mod",
                               full_season=False, drop_weisfluh_joch=True,
                               save_dir=None, filename="hnw_swe_validation_combined.png"):
    """
    Plot HNW and SWE validation side-by-side in one figure.

    Parameters
    ----------
    hnw_df : pd.DataFrame  — must contain hnw_obs_col, hnw_mod_col, station, time
    swe_df : pd.DataFrame  — must contain swe_obs_col, swe_mod_col, station, time
    model_name : str       — shown as figure suptitle
    params : dict, optional
        SnowToSwe parameters to annotate in the figure, e.g.
        dict(rho_max=381.203, rho_null=106.832, c_ov=0.00055347,
             k_ov=0.403141, k=0.0272148, tau=0.0222341, eta_null=8.65803e6)
    full_season : bool     — if False, restrict to Nov–Apr
    drop_weisfluh_joch : bool
    save_dir : str or Path, optional
    filename : str

    Returns
    -------
    dict with keys "HNW" and "SWE", each a stats dict
    (RMSE, Bias, Rel_BIAS, R2, N).
    """

    # ── HNW preparation ───────────────────────────────────────────────────────
    hnw = hnw_df.copy()
    if drop_weisfluh_joch:
        hnw = hnw[hnw["station"] != "Weisfluh_Joch"]
    hnw = _filter_season(hnw, full_season)
    hnw = hnw.dropna(subset=[hnw_obs_col, hnw_mod_col])
    hnw = hnw[hnw[hnw_obs_col] >= 0]
    hnw = hnw[np.isfinite(hnw[hnw_obs_col]) & np.isfinite(hnw[hnw_mod_col])]

    x_hnw = hnw[hnw_obs_col].values
    y_hnw = hnw[hnw_mod_col].values
    stats_hnw = _calculate_metrics(x_hnw, y_hnw)
    stats_hnw["N"] = len(hnw)

    # ── SWE preparation ───────────────────────────────────────────────────────
    swe = swe_df.copy()
    if drop_weisfluh_joch:
        swe = swe[swe["station"] != "Weisfluh_Joch"]
    swe = _filter_season(swe, full_season=True)   # SWE uses full year
    swe = swe.dropna(subset=[swe_obs_col, swe_mod_col])
    swe = swe[swe[swe_obs_col] >= 0]
    swe = swe[np.isfinite(swe[swe_obs_col]) & np.isfinite(swe[swe_mod_col])]

    x_swe = swe[swe_obs_col].values
    y_swe = swe[swe_mod_col].values
    stats_swe = _calculate_metrics(x_swe, y_swe)
    stats_swe["N"] = len(swe)

    print("HNW stats:", stats_hnw)
    print("SWE stats:", stats_swe)

    # ── Build params annotation string ────────────────────────────────────────
    param_str = ""
    if params is not None:
        parts = [f"{k}={v:.6g}" for k, v in params.items()]
        param_str = "  |  ".join(parts)

    # ── Figure ────────────────────────────────────────────────────────────────
    fig, axes = plt.subplots(1, 2, figsize=(16, 7))

    _plot_validation_ax(
        ax=axes[0], x=x_hnw, y=y_hnw, stats=stats_hnw,
        title="HNW validation",
        lim=[0, 100],
        xlabel="Observed HNW (mm)",
        ylabel="Modeled HNW (mm)",
        fig=fig
    )

    _plot_validation_ax(
        ax=axes[1], x=x_swe, y=y_swe, stats=stats_swe,
        title="SWE validation",
        lim=[0, 1000],
        xlabel="Observed SWE (mm)",
        ylabel="Modeled SWE (mm)",
        fig=fig
    )

    suptitle = model_name
    if param_str:
        suptitle += f"\n{param_str}"

    fig.suptitle(suptitle, fontsize=13, y=1.01)
    plt.tight_layout()

    if save_dir is not None:
        os.makedirs(save_dir, exist_ok=True)
        save_path = os.path.join(save_dir, filename)
        fig.savefig(save_path, dpi=300, bbox_inches="tight")
        print(f"Plot saved to: {save_path}")

    plt.show()

    return {"HNW": stats_hnw, "SWE": stats_swe}

File: HNW_validation/Archieve/HNW_validation_helper.R
# =============================================================================
# HNW_validation_helper.R
# R equivalent of HNW_validation_helper.py
#
# Provides three internal helper functions and two public validation functions:
#   validate_hnw_mag25()  – density scatter for HNW  (lim 0–100 mm)
#   validate_swe_mag25()  – density scatter for SWE  (lim 0–1000 mm)
#
# Dependencies: ggplot2, lubridate
# =============================================================================

library(ggplot2)
library(lubridate)


# ── Internal helpers (prefixed with . to signal private use) ──────────────────

#' Filter a data frame to the snow season (Nov 1 – Apr 30).
#' Passes through unchanged when full_season = TRUE.
#' Equivalent to _filter_season() in Python.
.filter_season <- function(df, full_season = FALSE) {
  df$time <- as.Date(df$time)
  if (full_season) return(df)
  m <- lubridate::month(df$time)
  df[m >= 11 | m <= 4, ]
}


#' Calculate standard validation metrics.
#' x = observed values, y = modelled values.
#' Returns a named list: RMSE, Bias, Rel_BIAS (relative / percent bias), R2.
#' Equivalent to _calculate_metrics() in Python.
.calc_metrics <- function(x, y) {
  res    <- y - x
  rmse   <- sqrt(mean(res^2))
  bias   <- mean(res)
  pbias  <- if (sum(x) != 0) sum(res) / sum(x) else NA_real_   # Rel_BIAS
  ss_res <- sum((x - y)^2)
  ss_tot <- sum((x - mean(x))^2)
  r2     <- if (ss_tot != 0) 1 - ss_res / ss_tot else NA_real_
  list(RMSE = rmse, Bias = bias, Rel_BIAS = pbias, R2 = r2)
}


#' Build a 2-D density (histogram) validation plot with log-scaled colour.
#' Reproduces matplotlib's hist2d + LogNorm + viridis used in Python.
#'
#' @param x          Observed values (numeric vector).
#' @param y          Modelled values (numeric vector, same length as x).
#' @param stats      Named list from .calc_metrics().
#' @param title      Plot title / model name.
#' @param lim        Length-2 numeric: shared axis limits c(lo, hi).
#' @param xlabel     x-axis label string.
#' @param ylabel     y-axis label string.
#' @return A ggplot2 object.
.plot_validation <- function(x, y, stats, title, lim, xlabel, ylabel) {

  # vmax: upper colour-scale limit = max(1, n / 10)  — mirrors Python's vmax
  vmax <- max(1, length(x) / 10)

  # Colour-bar ticks: powers of 10 that fit ≤ vmax, then always append vmax
  brk_cands <- c(1, 10, 100, 1000, 10000)
  brk       <- brk_cands[brk_cands <= vmax]
  if (!length(brk)) brk <- 1
  if (tail(brk, 1) != vmax) brk <- c(brk, vmax)
  brk_labels <- c(as.character(as.integer(head(brk, -1))),
                  as.character(as.integer(vmax)))

  # 5 evenly spaced axis ticks (mirrors np.linspace(lim[0], lim[1], 5))
  axis_brk <- seq(lim[1], lim[2], length.out = 5)

  # Stats annotation text placed at axes coordinates (0.03, 0.97) = top-left
  label_text <- sprintf(
    "R\u00b2: %.2f\nBias: %.2f\nRMSE: %.1f\nRel_BIAS: %.1f%%",
    stats$R2, stats$Bias, stats$RMSE, stats$Rel_BIAS * 100
  )

  ggplot(data.frame(x = x, y = y), aes(x, y)) +

    # 2-D histogram, log-scaled fill — matches hist2d(..., norm=LogNorm, cmap="viridis")
    stat_bin2d(bins = 50, aes(fill = after_stat(count))) +
    scale_fill_viridis_c(
      name   = "Number of observations",
      trans  = "log10",
      limits = c(1, vmax),
      breaks = brk,
      labels = brk_labels
    ) +

    # 1:1 reference line — matches plt.plot(lim, lim, "--", color="gray")
    geom_abline(linetype = "dashed", colour = "grey50", linewidth = 0.8) +

    # Axis limits and ticks — matches plt.xticks / plt.yticks + plt.xlim / plt.ylim
    scale_x_continuous(limits = lim, breaks = axis_brk) +
    scale_y_continuous(limits = lim, breaks = axis_brk) +

    labs(title = title, x = xlabel, y = ylabel) +

    # Stats annotation box — matches ax.text(0.03, 0.97, ..., va="top")
    annotate("label",
             x     = lim[1] + diff(lim) * 0.03,
             y     = lim[2] - diff(lim) * 0.03,
             label = label_text,
             hjust = 0, vjust = 1,
             size  = 4,
             fill  = "white", alpha = 0.8, label.size = 0.3) +

    # Minimal theme, no grid — matches plt.grid(False)
    theme_minimal(base_size = 13) +
    theme(
      panel.grid  = element_blank(),
      plot.title  = element_text(size = 15)
    )
}


# ── Public validation functions ───────────────────────────────────────────────

#' Validate modelled HNW against observations from the Mag25 dataset.
#' Exact R equivalent of validate_hnw_mag25() in HNW_validation_helper.py.
#'
#' @param df         Data frame with columns: time, <obs_col>, <mod_col>.
#' @param model_name String used as the plot title.
#' @param obs_col    Name of the observed HNW column  (default "HNW_obs").
#' @param mod_col    Name of the modelled HNW column  (default "HNW_mod").
#' @param save_dir   Directory for saving the PNG; NULL = do not save.
#' @param filename   PNG filename (used when save_dir is not NULL).
#' @param full_season If FALSE (default) restrict to the Nov–Apr snow season.
#' @return Named list of metrics (invisibly).
validate_hnw_mag25 <- function(df,
                               model_name,
                               obs_col     = "HNW_obs",
                               mod_col     = "HNW_mod",
                               save_dir    = NULL,
                               filename    = "hnw_validation.png",
                               full_season = FALSE) {

  df <- .filter_season(df, full_season)

  # Drop NA rows, keep only non-negative observations, remove non-finite values
  # Caution: >= 0 (not > 0) is intentional — matches the Python comment exactly
  df_valid <- df[!is.na(df[[obs_col]]) & !is.na(df[[mod_col]]), ]
  df_valid <- df_valid[df_valid[[obs_col]] >= 0, ]
  df_valid <- df_valid[is.finite(df_valid[[obs_col]]) & is.finite(df_valid[[mod_col]]), ]

  x <- df_valid[[obs_col]]
  y <- df_valid[[mod_col]]

  stats   <- .calc_metrics(x, y)
  stats$N <- nrow(df_valid)
  print(stats)

  p <- .plot_validation(x, y, stats,
                        title  = model_name,
                        lim    = c(0, 100),
                        xlabel = "Observed HNW (mm)",
                        ylabel = "Modeled HNW (mm)")

  if (!is.null(save_dir)) {
    dir.create(save_dir, showWarnings = FALSE, recursive = TRUE)
    save_path <- file.path(save_dir, filename)
    ggsave(save_path, plot = p, dpi = 300, width = 8, height = 7,
           units = "in", bg = "white")
    cat(sprintf("Plot saved to: %s\n", save_path))
  }

  print(p)
  invisible(stats)
}


#' Validate modelled SWE against observations from the Mag25 dataset.
#' Exact R equivalent of validate_swe_mag25() in HNW_validation_helper.py.
#'
#' @inheritParams validate_hnw_mag25
#' @param obs_col Name of the observed SWE column (default "SWE_obs").
#' @param mod_col Name of the modelled SWE column (default "SWE_mod").
validate_swe_mag25 <- function(df,
                               model_name,
                               obs_col     = "SWE_obs",
                               mod_col     = "SWE_mod",
                               save_dir    = NULL,
                               filename    = "swe_validation.png",
                               full_season = FALSE) {

  df <- .filter_season(df, full_season)

  df_valid <- df[!is.na(df[[obs_col]]) & !is.na(df[[mod_col]]), ]
  df_valid <- df_valid[df_valid[[obs_col]] >= 0, ]

  # Mirrors Python's explicit print before the finite filter
  cat(sprintf("Number of valid observations after filtering: %d\n", nrow(df_valid)))

  df_valid <- df_valid[is.finite(df_valid[[obs_col]]) & is.finite(df_valid[[mod_col]]), ]

  x <- df_valid[[obs_col]]
  y <- df_valid[[mod_col]]

  stats   <- .calc_metrics(x, y)
  stats$N <- nrow(df_valid)
  print(stats)

  p <- .plot_validation(x, y, stats,
                        title  = model_name,
                        lim    = c(0, 1000),
                        xlabel = "Observed SWE (mm)",
                        ylabel = "Modeled SWE (mm)")

  if (!is.null(save_dir)) {
    dir.create(save_dir, showWarnings = FALSE, recursive = TRUE)
    save_path <- file.path(save_dir, filename)
    ggsave(save_path, plot = p, dpi = 300, width = 8, height = 7,
           units = "in", bg = "white")
    cat(sprintf("Plot saved to: %s\n", save_path))
  }

  print(p)
  invisible(stats)
}
